kani-driver/src/cbmc_output_parser.rs

```diff
@@ -0,0 +1,991 @@
+    }
+}
+
+/// Struct that represents a property.
+///
+/// Note: `reach` is not part of the parsed data, but it's useful to annotate
+/// its reachability status.
+#[derive(Clone, Debug, Deserialize)]
+pub struct Property {
+    pub description: String,
+    pub property: String,
+    #[serde(rename = "sourceLocation")]
+    pub source_location: SourceLocation,
+    pub status: CheckStatus,
+    pub reach: Option<CheckStatus>,
+    pub trace: Option<Vec<TraceItem>>,
+}
+
+/// Struct that represents a source location.
+///
+/// Source locations may be completely empty, which is why
+/// all members are optional.
+#[derive(Clone, Debug, Deserialize)]
+pub struct SourceLocation {
+    pub column: Option<String>,
+    pub file: Option<String>,
+    pub function: Option<String>,
+    pub line: Option<String>,
+}
+
+impl SourceLocation {
+    /// Determines if fundamental parts of a source location are missing.
+    fn is_missing(&self) -> bool {
+        self.file.is_none() && self.function.is_none()
+    }
+}
+
+/// `Display` implement for `SourceLocation`.
+///
+/// This is used to format source locations for individual checks. But source
+/// locations may be printed in a different way in other places (e.g., in the
+/// "Failed Checks" summary at the end).
+///
+/// Source locations formatted this way will look like:
+/// `<file>:<line>:<column> in function <function>`
+/// if all attributes were specified. Otherwise, we:
+///  * Omit `in function <function>` if the function isn't specified.
+///  * Use `Unknown file` instead of `<file>:<line>:<column>` if the file isn't
+///    specified.
+///  * Lines and columns are only formatted if they were specified and preceding
+///    attribute was formatted.
+impl std::fmt::Display for SourceLocation {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut fmt_str = String::new();
+        if let Some(file) = self.file.clone() {
+            let file_str = filepath(file);
+            fmt_str.push_str(file_str.as_str());
+            if let Some(line) = self.line.clone() {
+                let line_str = format!(":{}", line);
+                fmt_str.push_str(line_str.as_str());
+                if let Some(column) = self.column.clone() {
+                    let column_str = format!(":{}", column);
+                    fmt_str.push_str(column_str.as_str());
+                }
+            }
+        } else {
+            fmt_str.push_str("Unknown file");
+        }
+        if let Some(function) = self.function.clone() {
+            let fun_str = format!(" in function {}", function);
+            fmt_str.push_str(fun_str.as_str());
+        }
+
+        write! {f, "{}", fmt_str}
+    }
+}
+
+/// Returns a path relative to the current working directory.
+fn filepath(file: String) -> String {
+    let file_path = PathBuf::from(file.clone());
```

Comment:
Can't this be an `.as_ref()` or `.into()` instead of `.clone()`?

Reply:
Yes, but then I need to add type annotations as follows:
```
help: try using a fully qualified path to specify the expected types
    |
295 |     let file_path = PathBuf::from(<std::string::String as AsRef<T>>::as_ref(&file));
    |                                   +++++++++++++++++++++++++++++++++++++++++++    ~

For more information about this error, try `rustc --explain E0283`.
```

Doesn't look so readable to me. What do you think?

Reply:
Leave it as is. Apparently the types are more complicated than what I inferred in my head.